diff --git a/src/view/home/HomeBox.py b/src/view/home/HomeBox.py
index 39c24a4..f7f698a 100644
--- a/src/view/home/HomeBox.py
+++ b/src/view/home/HomeBox.py
@@ -139,7 +139,7 @@ class HomeToolbar(gtk.Toolbar):
         self._add_separator(expand=True)
 
         ring_button = RadioToolButton(named_icon='view-radial', group=None)
-        ring_button.props.label = _('Ring view')
+        ring_button.props.tooltip = _('Ring view')
         ring_button.props.accelerator = _('<Ctrl>R')
         ring_button.connect('toggled', self.__view_button_toggled_cb, _RING_VIEW)
         self.insert(ring_button, -1)
@@ -147,7 +147,7 @@ class HomeToolbar(gtk.Toolbar):
 
         list_button = RadioToolButton(named_icon='view-list')
         list_button.props.group = ring_button
-        list_button.props.label = _('List view')
+        list_button.props.tooltip = _('List view')
         list_button.props.accelerator = _('<Ctrl>L')
         list_button.connect('toggled', self.__view_button_toggled_cb, _LIST_VIEW)
         self.insert(list_button, -1)
diff --git a/src/view/home/HomeWindow.py b/src/view/home/HomeWindow.py
index 607272e..9cb7c77 100644
--- a/src/view/home/HomeWindow.py
+++ b/src/view/home/HomeWindow.py
@@ -36,7 +36,9 @@ class HomeWindow(gtk.Window):
     def __init__(self):
         gtk.Window.__init__(self)
 
-        self.add_accel_group(gtk.AccelGroup())
+        accel_group = gtk.AccelGroup()
+        self.set_data('sugar-accel-group', accel_group)
+        self.add_accel_group(accel_group)
 
         self._active = False
         self._level = ShellModel.ZOOM_HOME
diff --git a/src/view/home/MeshBox.py b/src/view/home/MeshBox.py
index b166edc..427e08a 100644
--- a/src/view/home/MeshBox.py
+++ b/src/view/home/MeshBox.py
@@ -159,7 +159,7 @@ class MeshDeviceView(CanvasPulsingIcon):
         if not channel in [1, 6, 11]:
             raise ValueError("Invalid channel %d" % channel)
 
-        PulsingIcon.__init__(self, size=style.STANDARD_ICON_SIZE,
+        CanvasPulsingIcon.__init__(self, size=style.STANDARD_ICON_SIZE,
                              icon_name=_MESH_ICON_NAME, cache=True)
 
         self._nm_device = nm_device
diff --git a/src/view/pulsingicon.py b/src/view/pulsingicon.py
index 0a4a825..e7465a5 100644
--- a/src/view/pulsingicon.py
+++ b/src/view/pulsingicon.py
@@ -185,8 +185,9 @@ class CanvasPulsingIcon(CanvasIcon):
         self._pulse_hid = gobject.timeout_add(_INTERVAL, self.__pulse_cb)
 
     def _stop_pulsing(self):
-        gobject.source_remove(self._pulse_hid)
-        self._pulse_hid = None
+        if self._pulse_hid is not None:
+            gobject.source_remove(self._pulse_hid)
+            self._pulse_hid = None
         self.props.xo_color = self._base_color
 
     def __pulse_cb(self):
