--- sugar-0.79.0.orig/src/hardware/keydialog.py
+++ sugar-0.79.0/src/hardware/keydialog.py
@@ -266,7 +266,7 @@
         elif len(key) >= 8 and len(key) <= 63:
             # passphrase
             import commands
-            command = "/usr/sbin/wpa_passphrase '%s' '%s'" % (ssid, key)
+            command = "/usr/bin/wpa_passphrase '%s' '%s'" % (ssid, key)
             (s, o) = commands.getstatusoutput(command)
             if s != 0:
                 raise RuntimeError("Error hashing passphrase: %s" % o)
