--- a/src/jarabe/desktop/keydialog.py
+++ b/src/jarabe/desktop/keydialog.py
@@ -239,7 +239,7 @@
         elif len(key) >= 8 and len(key) <= 63:
             # passphrase
             from subprocess import Popen, PIPE
-            p = Popen(['/usr/sbin/wpa_passphrase', ssid, key], stdout=PIPE)
+            p = Popen(['/usr/bin/wpa_passphrase', ssid, key], stdout=PIPE)
             for line in p.stdout:
                 if line.strip().startswith("psk="):
                     real_key = line.strip()[4:]
