diff --git a/configure b/configure
index 1c17718..7076798 100755
--- a/configure
+++ b/configure
@@ -2641,7 +2641,7 @@ fi
 
 # Setup GETTEXT
 #
-ALL_LINGUAS="af am ar ay bg bn bn_IN ca de dz el en es fa fa_AF ff fr gu ha hi ht ig is it ja km ko mk ml mn mr mvo nb ne nl pa pap pis pl ps pt pt_BR qu ro ru rw sd si sl te th tpi tr ur vi yo zh_CN zh_TW"
+ALL_LINGUAS="af am ar ay bg bn bn_IN ca da de dz el en es fa fa_AF ff fr gu ha hi ht ig is it ja km ko mk ml mn mr mvo nb ne nl pa pap pis pl ps pt pt_BR qu ro ru rw sd si sl te th tpi tr ur vi yo zh_CN zh_TW"
 GETTEXT_PACKAGE=sugar
 DEPDIR="${am__leading_dot}deps"
 
