--- a/bin/sugar-emulator
+++ b/bin/sugar-emulator
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-if [ "$(id -u)" -eq 0 -o "$(id -ru)" -eq 0 ] ; then
+if [ "$(id -u)" -eq 0 ] || [ "$(id -ru)" -eq 0 ] ; then
 	echo Refusing to run as root.
 	exit 3
 fi
--- a/bin/sugar.in
+++ b/bin/sugar.in
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-if [ "$(id -u)" -eq 0 -o "$(id -ru)" -eq 0 ] ; then
+if [ "$(id -u)" -eq 0 ] || [ "$(id -ru)" -eq 0 ] ; then
 	echo Refusing to run as root.
 	exit 3
 fi
@@ -60,7 +60,7 @@
 export LANGUAGE="${LANGUAGE:-${LANG}}"
 
 # Set Sugar's telepathy accounts directory
-export MC_ACCOUNT_DIR=$HOME/.sugar/$SUGAR_PROFILE/accounts
+export MC_ACCOUNT_DIR="$HOME/.sugar/$SUGAR_PROFILE/accounts"
 
 # Source language settings and debug definitions
 if [ -f ~/.i18n ]; then
@@ -71,6 +71,6 @@
 fi
 
 echo Xcursor.theme: sugar | xrdb -merge
-metacity --no-force-fullscreen -d $DISPLAY &
+metacity --no-force-fullscreen -d "$DISPLAY" &
 
 exec sugar-session
